### ISSUE TITLE
nan appearing on False fork of tf.where propgates to e

### ISSUE THREAD

**Issue by cottrell** [tensorflow/tensorflow#27684]
https://github.com/tensorflow/tensorflow/issues/20091

Same issue, but that one was closed. The fix is to protect every single argument against the fork.

```python
import tensorflow as tf
import numpy as np
x = tf.convert_to_tensor(np.random.randn(100).astype(np.float32))
alpha = tf.Variable(1, name='asdf', dtype=tf.float32, trainable=True)
with tf.GradientTape(persistent=True) as t:
    safe_x = tf.where(x > 0, x, tf.ones_like(x))
    # x = safe_x # this fixes it
    s = tf.where(x >= 0, alpha * tf.math.log(x), tf.zeros_like(x))
gradients = t.gradient(s, [alpha])


In [326]: reload(d); d.gradients
Out[326]: [<tf.Tensor: id=2832835, shape=(), dtype=float32, numpy=-35.19606>]

In [327]: reload(d); d.gradients
Out[327]: [<tf.Tensor: id=2832875, shape=(), dtype=float32, numpy=nan>]

In [346]: sys.version
Out[346]: '3.7.3 (default, Mar 27 2019, 16:54:48) \n[Clang 4.0.1 (tags/RELEASE_401/final)]'

In [347]: platform.platform()
Out[347]: 'Darwin-17.7.0-x86_64-i386-64bit'

In [348]: tf.__version__
Out[348]: '2.0.0-alpha0'

```

**Comment by ymodak** [comment]
@cottrell Thanks for trying TF 2.0 Alpha. The code snippet you provided looks incomplete. Can you please provide minimal code snippet to reproduce the issue reported here? Thanks!

**Comment by cottrell** [comment]
I was missing a "np."float32 but added that. Otherwise I think it is complete (I just copy pasted it and it ran). I think the confusing thing could be I saved it in a file called "d.py" or something like and was imp.reloading it from the REPL to show the changes.

**Comment by cottrell** [comment]
To be clear, this one works (with the bit uncommented).

```
In [130]: import tensorflow as tf
     ...: import numpy as np
     ...: x = tf.convert_to_tensor(np.random.randn(100).astype(np.float32))
     ...: alpha = tf.Variable(1, name='asdf', dtype=tf.float32, trainable=True)
     ...: with tf.GradientTape(persistent=True) as t:
     ...:     safe_x = tf.where(x > 0, x, tf.ones_like(x))
     ...:     x = safe_x # this fixes it
     ...:     s = tf.where(x >= 0, alpha * tf.math.log(x), tf.zeros_like(x))
     ...: gradients = t.gradient(s, [alpha])
```

**Comment by ebrevdo** [comment]
Does this replicate in TF2.0 nightly or has it been fixed?

**Comment by ebrevdo** [comment]
Up until about a month(?) ago, your solution was the only solution.  But the nightly branch now implements a safe gradient operation that should zero out during backprop.  But maybe not directly for the gradient of tertiary where.  @rmlarsen maybe we need the mul_no_nan in the SelectOp gradient as well?

**Comment by ebrevdo** [comment]
I believe we decided that the "automatically safe solution" hides some true error cases by converting gradients of real nans to 0s.

So for now we've disabled the automatic zero-ing out of grad(x)*dy where dy  == 0.

Instead we should update the docstring of tf.math.log to suggest using tf.math.xlogy, and the division operator to use tf.math.xdivy if y is expected to be 0.

@rmlarsen @ezhulenev should we do an automatic grappler rewrite for x*log(y) -> xlogy(x,y)?  i think it would lead to tf.function mode being inconsistent with eager mode.  Perhaps that's ok?

**Comment by alextp** [comment]
I recommend reading https://stackoverflow.com/a/42497444/433287 for a description of the most common workaround to the nan-where problem.

Eugene, I think the automatic rewrite is a little dangerous because not every single case of x log y wants to compute the (cross) entropy.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2 and TF-nightly i.e. v2.4.0-dev20200722. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/151e1acbe8b0513458f9c7da13187a2f/27684.ipynb#scrollTo=UGHJaUo0xHGm). Thanks!

**Comment by sushreebarsa** [comment]
@cottrell Could you please try on latest stable version of tf 2.5 or 2.4.1 and let us know if this is still an issue.Thanks!

**Comment by cottrell** [comment]
Looks like it still persists (am on a different machine now, Linux with GPU but I think that shouldn't matter for this problem).

```
In [1]: import tensorflow as tf 
   ...: import numpy as np 
   ...: x = tf.convert_to_tensor(np.random.randn(100).astype(np.float32)) 
   ...: alpha = tf.Variable(1, name='asdf', dtype=tf.float32, trainable=True) 
   ...: with tf.GradientTape(persistent=True) as t: 
   ...:     safe_x = tf.where(x > 0, x, tf.ones_like(x)) 
   ...:     # x = safe_x # this fixes it 
   ...:     s = tf.where(x >= 0, alpha * tf.math.log(x), tf.zeros_like(x)) 
   ...: t.gradient(s, [alpha])                                                                                                                                                                                     

Out[1]: [<tf.Tensor: shape=(), dtype=float32, numpy=nan>]

In [2]: import tensorflow as tf 
   ...: import numpy as np 
   ...: x = tf.convert_to_tensor(np.random.randn(100).astype(np.float32)) 
   ...: alpha = tf.Variable(1, name='asdf', dtype=tf.float32, trainable=True) 
   ...: with tf.GradientTape(persistent=True) as t: 
   ...:     safe_x = tf.where(x > 0, x, tf.ones_like(x)) 
   ...:     x = safe_x # this fixes it 
   ...:     s = tf.where(x >= 0, alpha * tf.math.log(x), tf.zeros_like(x)) 
   ...: t.gradient(s, [alpha])                                                                                                                                                                                     
Out[2]: [<tf.Tensor: shape=(), dtype=float32, numpy=-38.554565>]

In [3]: import sys, platform                                                                                                                                                                                       

In [4]: sys.version                                                                                                                                                                                                
Out[4]: '3.8.3 (default, May 19 2020, 18:47:26) \n[GCC 7.3.0]'

In [5]: platform.platform()                                                                                                                                                                                        
Out[5]: 'Linux-5.4.0-65-generic-x86_64-with-glibc2.10'

In [6]: tf.__version__                                                                                                                                                                                             
Out[6]: '2.5.0'
```

**Comment by sanatmpa1** [comment]
Issue still persists in `TF 2.6.0` and `TF-nightly`. Please find the [gist here](https://colab.research.google.com/gist/sanatmpa1/9be46297dafef158ba1e2ffb0cd96524/27684.ipynb). Thanks!

**Comment by ebrevdo** [comment]
We rolled back the "safe rewrite" because, as @alextp mentioned, the assumption doesn't always hold.

In other words, this behavior is backprop working as intended!  Trying to automagically fix it breaks folks who rely on the non-safe multiply during backprop.  Closing.
